bundle/edu.gemini.pot/src/test/scala/edu/gemini/spModel/target/env/GuideGroupSpec.scala

```diff
@@ -177,90 +175,92 @@ class GuideGroupSpec extends Specification with ScalaCheck with Arbitraries {
+          gpOut.forall(gp => !g2.contains(gp))
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getPrimaryReferencedGuiders" should {
+    "contain a guider iff it has a primary target" in
+      forAll { g: GuideGroup =>
+        g.getAll.asScalaList.collect {
+          case gpt if gpt.getPrimary.isDefined => gpt.getGuider
+        }.toSet == g.getPrimaryReferencedGuiders.asScala.toSet
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getReferencedGuiders" should {
+    "contain a guider iff it has associated targets" in
+      forAll { g: GuideGroup =>
+        g.getAll.asScalaList.map(_.getGuider).toSet == g.getReferencedGuiders.asScala.toSet
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup removeTarget" should {
+    "remove an existing target added to all guiders and preserve the order of the original targets" in
+      forAll { (g: GuideGroup, target: SPTarget) =>
+        val oldGpts = g.getAll
+        val newGpts = g.getAll.asScalaList.map(gt => gt.update(OptionsList.UpdateOps.append(target))).asImList
+        val remGpts = g.setAll(newGpts).removeTarget(target).getAll
+        remGpts == oldGpts
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getAllMatching" should {
+    "partition the collections of guide probe targets by guide probe type" in
+      forAll { g: GuideGroup =>
+        val ptns = g.getReferencedGuiders.asScala.map(gp => g.getAllMatching(gp.getType).asScalaList.toSet).toSet
+        (ptns.toList.map(_.length).sum == g.getAll.size) && ptns.forall(p => ptns.forall(q => p == q || p.intersect(q).isEmpty))
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup cloneTargets" should {
+    "create a new GuideGroup where the underlying ITargets are equal for each guide probe and in the same order" in
+      forAll { g: GuideGroup =>
+        val cg = g.cloneTargets
+        val gt = g.getAll.asScalaList
+        val cgt = cg.getAll.asScalaList
+        (gt.length == cgt.length) && gt.zip(cgt).forall {
+          case (t1, t2) => t1.getGuider == t2.getGuider && t1.getTargets.asScala.map(_.getTarget) == t2.getTargets.asScala.map(_.getTarget)
+        }
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
-      }
+  "GuideGroup getAllContaining" should {
+    "return a subset of guide probe targets containing a specific target for manual groups" in
+    forAll { (g: GuideGroup, t: SPTarget) =>
+      val newGpt = g.getAll.asScalaList.zipWithIndex.map { case (gpt, idx) => if (idx % 2 == 0) gpt else gpt.update(OptionsList.UpdateOps.append(t)) }
+      g.setAll(newGpt.asImList).getAllContaining(t).asScalaList == (g.grp match {
+        case _: ManualGroup    => newGpt.zipWithIndex.collect { case (gpt, idx) if idx % 2 == 1 => gpt }
+        case _: AutomaticGroup => Nil
+      })
+    }
```

Comment:
I think that this is the last thing I have to look at, so I'll do that tomorrow.
Let me know if you have any further notes!